File: scripts/taku_publisher/discovery.py
from __future__ import annotations

import hashlib
import json
import os
import re
from pathlib import Path
from typing import Any

from .constants import (
    EXCLUDED_DIR_NAMES,
    MAX_DISCOVERY_DEPTH,
    SUPPORTED_TYPES,
    UNAVAILABLE_PUBLISH_TYPES,
)
from .util import PublisherError, normalized_relative


PLUGIN_MANIFESTS = (".codex-plugin/plugin.json", "taku.plugin.json", "plugin.json")
AGENT_FILE_NAMES = {"agent.md", "agents.md", "agent.json", "agent.yaml", "agent.yml"}
ACTION_PARENT_NAMES = {"action", "actions", "command", "commands", "workflow", "workflows"}
AGENT_PARENT_NAMES = {"agent", "agents"}


def discover_units(workspace: Path, explicit_source: Path | None = None) -> list[dict[str, Any]]:
    workspace = workspace.expanduser().resolve()
    if not workspace.is_dir():
        raise PublisherError("Workspace must be an existing directory.", code="invalid_workspace")
    if explicit_source is not None:
        raw_source = explicit_source.expanduser()
        if raw_source.is_symlink():
            raise PublisherError("An explicit publish source cannot be a symlink.", code="unsafe_source")
        source = raw_source.resolve()
        if not source.exists():
            raise PublisherError("Explicit source does not exist.", code="invalid_source")
        return _discover_explicit(source, workspace)

    candidates: list[dict[str, Any]] = []
    seen: set[tuple[str, str]] = set()
    for current_root, directories, files in os.walk(workspace, topdown=True, followlinks=False):
        current = Path(current_root)
        depth = len(current.relative_to(workspace).parts)
        directories[:] = sorted(
            directory
            for directory in directories
            if directory.lower() not in EXCLUDED_DIR_NAMES
            and not (current / directory).is_symlink()
            and depth < MAX_DISCOVERY_DEPTH
        )
        files = sorted(files)

        plugin_manifest = _plugin_manifest(current)
        if plugin_manifest:
            _append_candidate(candidates, seen, _candidate("plugin", current, workspace, plugin_manifest))
        if "SKILL.md" in files:
            _append_candidate(candidates, seen, _candidate("skill", current, workspace, current / "SKILL.md"))

        for file_name in files:
            file_path = current / file_name
            lower = file_name.lower()
            parent = current.name.lower()
            if lower in AGENT_FILE_NAMES or (parent in AGENT_PARENT_NAMES and file_path.suffix.lower() == ".md"):
                if file_path.suffix.lower() in {".md", ".json", ".yaml", ".yml"}:
                    _append_candidate(candidates, seen, _candidate("agent", file_path, workspace, file_path))
            if parent in ACTION_PARENT_NAMES and file_path.suffix.lower() in {".md", ".json", ".yaml", ".yml"}:
                _append_candidate(candidates, seen, _candidate("action", file_path, workspace, file_path))

    return sorted(candidates, key=lambda entry: (entry["type"], entry["relative_path"], entry["id"]))


def inspect_selected_unit(source: Path, unit_type: str, workspace: Path) -> dict[str, Any]:
    assert_publish_type_available(unit_type)
    raw_source = source.expanduser()
    if raw_source.is_symlink():
        raise PublisherError("A selected publish unit cannot be a symlink.", code="unsafe_source")
    source = raw_source.resolve()
    workspace = workspace.expanduser().resolve()
    if not source.exists():
        raise PublisherError("Selected source does not exist.", code="invalid_source")

    entrypoint: Path
    if unit_type == "skill":
        root = source.parent if source.is_file() else source
        entrypoint = root / "SKILL.md"
        if not entrypoint.is_file():
            raise PublisherError("A skill source must contain SKILL.md.", code="invalid_skill")
        source = root
    elif unit_type == "plugin":
        if not source.is_dir():
            raise PublisherError("A plugin source must be a directory.", code="invalid_plugin")
        entrypoint = _plugin_manifest(source) or Path()
        if not entrypoint or not entrypoint.is_file():
            raise PublisherError(
                "A plugin source must contain .codex-plugin/plugin.json, taku.plugin.json, or plugin.json.",
                code="invalid_plugin",
            )
    else:
        if source.is_file():
            entrypoint = source
        else:
            expected = [source / name for name in sorted(AGENT_FILE_NAMES)] if unit_type == "agent" else []
            entrypoint = next((path for path in expected if path.is_file()), Path())
            if not entrypoint:
                supported = sorted(
                    path for path in source.iterdir()
                    if path.is_file() and path.suffix.lower() in {".md", ".json", ".yaml", ".yml"}
                )
                entrypoint = supported[0] if supported else Path()
        if not entrypoint or not entrypoint.is_file():
            raise PublisherError(f"The selected {unit_type} has no supported definition file.", code=f"invalid_{unit_type}")

    candidate = _candidate(unit_type, source, workspace, entrypoint)
    if unit_type == "plugin":
        candidate["children"] = discover_plugin_children(source)
    return candidate


def assert_publish_type_available(unit_type: str) -> None:
    if unit_type in UNAVAILABLE_PUBLISH_TYPES:
        raise PublisherError(
            f"{unit_type} publishing is not available yet. Taku Publisher currently accepts Skills only.",
            code="publish_type_not_available",
        )
    if unit_type not in SUPPORTED_TYPES:
        raise PublisherError(f"Unsupported publish type: {unit_type}", code="unsupported_type")


def discover_plugin_children(plugin_root: Path) -> list[dict[str, Any]]:
    children: list[dict[str, Any]] = []
    for candidate in discover_units(plugin_root):
        if candidate["type"] == "plugin" and candidate["relative_path"] == ".":
            continue
        children.append({
            "id": candidate["id"],
            "type": candidate["type"],
            "name": candidate["name"],
            "relative_path": candidate["relative_path"],
        })
    return children


def _discover_explicit(source: Path, workspace: Path) -> list[dict[str, Any]]:
    candidates: list[dict[str, Any]] = []
    if source.is_dir():
        manifest = _plugin_manifest(source)
        if manifest:
            candidates.append(_candidate("plugin", source, workspace, manifest))
        if (source / "SKILL.md").is_file():
            candidates.append(_candidate("skill", source, workspace, source / "SKILL.md"))
        if not any(candidate["type"] in SUPPORTED_TYPES for candidate in candidates):
            candidates.extend(discover_units(source))
    else:
        lower = source.name.lower()
        parent = source.parent.name.lower()
        if lower == "skill.md":
            candidates.append(_candidate("skill", source.parent, workspace, source))
        if lower in AGENT_FILE_NAMES or parent in AGENT_PARENT_NAMES:
            candidates.append(_candidate("agent", source, workspace, source))
        if parent in ACTION_PARENT_NAMES or source.suffix.lower() in {".md", ".json", ".yaml", ".yml"}:
            candidates.append(_candidate("action", source, workspace, source))
    unique: dict[tuple[str, str], dict[str, Any]] = {}
    for candidate in candidates:
        unique[(candidate["type"], candidate["path"])] = candidate
    return sorted(
        (candidate for candidate in unique.values() if candidate["type"] in SUPPORTED_TYPES),
        key=lambda entry: (entry["type"], entry["relative_path"]),
    )


def _plugin_manifest(root: Path) -> Path | None:
    for relative in PLUGIN_MANIFESTS:
        candidate = root / relative
        if candidate.is_file():
            return candidate
    return None


def _candidate(unit_type: str, source: Path, workspace: Path, entrypoint: Path) -> dict[str, Any]:
    source = source.resolve()
    entrypoint = entrypoint.resolve()
    try:
        relative_path = normalized_relative(source, workspace)
    except ValueError:
        relative_path = source.name
    metadata = _metadata(entrypoint, source)
    identifier = hashlib.sha256(f"{unit_type}\0{relative_path}\0{entrypoint.name}".encode("utf-8")).hexdigest()[:20]
    return {
        "id": f"{unit_type}_{identifier}",
        "type": unit_type,
        "name": metadata["name"],
        "description": metadata.get("description", ""),
        "path": str(source),
        "relative_path": relative_path,
        "entrypoint": str(entrypoint),
        "entrypoint_relative": (
            entrypoint.relative_to(source).as_posix()
            if source.is_dir() and _is_relative_to(entrypoint, source)
            else entrypoint.name
        ),
    }


def _metadata(entrypoint: Path, source: Path) -> dict[str, str]:
    fallback = source.stem if source.is_file() else source.name
    if entrypoint.suffix.lower() == ".json":
        try:
            value = json.loads(entrypoint.read_text(encoding="utf-8"))
            if isinstance(value, dict):
                return {
                    "name": _text(value.get("name") or value.get("title")) or fallback,
                    "description": _text(value.get("description") or value.get("summary")),
                }
        except (OSError, json.JSONDecodeError, UnicodeDecodeError):
            pass
    try:
        text = entrypoint.read_text(encoding="utf-8")[:64_000]
    except (OSError, UnicodeDecodeError):
        return {"name": fallback, "description": ""}
    frontmatter = _frontmatter(text)
    heading = re.search(r"^#\s+(.+?)\s*$", text, re.MULTILINE)
    return {
        "name": _text(frontmatter.get("name") or frontmatter.get("title")) or (heading.group(1).strip() if heading else fallback),
        "description": _text(frontmatter.get("description") or frontmatter.get("summary")),
    }


def _frontmatter(text: str) -> dict[str, str]:
    if not text.startswith("---"):
        return {}
    end = text.find("\n---", 3)
    if end < 0:
        return {}
    output: dict[str, str] = {}
    for line in text[3:end].splitlines():
        match = re.match(r"^([A-Za-z0-9_-]+):\s*[\"']?(.*?)[\"']?\s*$", line)
        if match:
            output[match.group(1)] = match.group(2)
    return output


def _text(value: Any) -> str:
    return re.sub(r"\s+", " ", value).strip()[:500] if isinstance(value, str) else ""


def _append_candidate(
    candidates: list[dict[str, Any]],
    seen: set[tuple[str, str]],
    candidate: dict[str, Any],
) -> None:
    if candidate["type"] not in SUPPORTED_TYPES:
        return
    key = (candidate["type"], candidate["path"])
    if key not in seen:
        seen.add(key)
        candidates.append(candidate)


def _is_relative_to(path: Path, root: Path) -> bool:
    try:
        path.relative_to(root)
        return True
    except ValueError:
        return False

File: packages/publisher-runtime/src/discovery.ts
import { createHash } from 'node:crypto';
import * as fs from 'node:fs';
import * as fsp from 'node:fs/promises';
import * as path from 'node:path';

import {
  EXCLUDED_DIR_NAMES,
  MAX_DISCOVERY_DEPTH,
  SUPPORTED_TYPES,
  UNAVAILABLE_PUBLISH_TYPES,
} from './constants.js';
import type { JsonObject } from './types.js';
import { isWithin, normalizedRelative, PublisherError } from './util.js';

const PLUGIN_MANIFESTS = ['.codex-plugin/plugin.json', 'taku.plugin.json', 'plugin.json'];
const AGENT_FILE_NAMES = new Set(['agent.md', 'agents.md', 'agent.json', 'agent.yaml', 'agent.yml']);
const ACTION_PARENT_NAMES = new Set(['action', 'actions', 'command', 'commands', 'workflow', 'workflows']);
const AGENT_PARENT_NAMES = new Set(['agent', 'agents']);
const DEFINITION_SUFFIXES = new Set(['.md', '.json', '.yaml', '.yml']);
const HOST_METADATA_DIR_NAMES = new Set(['.codex-plugin', '.claude-plugin']);

export async function discoverUnits(workspaceInput: string, explicitSource?: string): Promise<JsonObject[]> {
  const workspace = await canonicalPath(workspaceInput);
  if (!(await isDirectory(workspace))) {
    throw new PublisherError('Workspace must be an existing directory.', 'invalid_workspace');
  }
  if (explicitSource) {
    const rawSource = path.resolve(explicitSource);
    if (await isSymlink(rawSource)) {
      throw new PublisherError('An explicit publish source cannot be a symlink.', 'unsafe_source');
    }
    if (!fs.existsSync(rawSource)) {
      throw new PublisherError('Explicit source does not exist.', 'invalid_source');
    }
    return discoverExplicit(await canonicalPath(rawSource), workspace);
  }

  const candidates: JsonObject[] = [];
  const seen = new Set<string>();
  await visitWorkspace(workspace, workspace, 0, async (current, files) => {
    const manifestPath = await pluginManifest(current);
    if (manifestPath) appendCandidate(candidates, seen, await candidate('plugin', current, workspace, manifestPath));
    if (files.includes('SKILL.md')) {
      appendCandidate(candidates, seen, await candidate('skill', current, workspace, path.join(current, 'SKILL.md')));
    }
    for (const fileName of files) {
      const filePath = path.join(current, fileName);
      const lower = fileName.toLowerCase();
      const parent = path.basename(current).toLowerCase();
      const suffix = path.extname(fileName).toLowerCase();
      if ((AGENT_FILE_NAMES.has(lower) || (AGENT_PARENT_NAMES.has(parent) && suffix === '.md'))
        && DEFINITION_SUFFIXES.has(suffix)) {
        appendCandidate(candidates, seen, await candidate('agent', filePath, workspace, filePath));
      }
      if (ACTION_PARENT_NAMES.has(parent) && DEFINITION_SUFFIXES.has(suffix)) {
        appendCandidate(candidates, seen, await candidate('action', filePath, workspace, filePath));
      }
    }
  });
  return candidates.sort(compareCandidates);
}

export async function inspectSelectedUnit(
  sourceInput: string,
  unitType: string,
  workspaceInput: string,
): Promise<JsonObject> {
  assertPublishTypeAvailable(unitType);
  let source = path.resolve(sourceInput);
  if (await isSymlink(source)) {
    throw new PublisherError('A selected publish unit cannot be a symlink.', 'unsafe_source');
  }
  if (!fs.existsSync(source)) {
    throw new PublisherError('Selected source does not exist.', 'invalid_source');
  }
  source = await canonicalPath(source);
  const workspace = await canonicalPath(workspaceInput);

  let entrypoint = '';
  if (unitType === 'skill') {
    const sourceStat = await fsp.stat(source);
    const root = sourceStat.isFile() ? path.dirname(source) : source;
    entrypoint = path.join(root, 'SKILL.md');
    if (!(await isFile(entrypoint))) {
      throw new PublisherError('A skill source must contain SKILL.md.', 'invalid_skill');
    }
    source = root;
  } else if (unitType === 'plugin') {
    if (!(await isDirectory(source))) {
      throw new PublisherError('A plugin source must be a directory.', 'invalid_plugin');
    }
    entrypoint = await pluginManifest(source) ?? '';
    if (!entrypoint) {
      throw new PublisherError(
        'A plugin source must contain .codex-plugin/plugin.json, taku.plugin.json, or plugin.json.',
        'invalid_plugin',
      );
    }
  } else {
    if (await isFile(source)) {
      entrypoint = source;
    } else {
      const entries = await fsp.readdir(source);
      if (unitType === 'agent') {
        entrypoint = entries
          .map((name) => path.join(source, name))
          .find((file) => AGENT_FILE_NAMES.has(path.basename(file).toLowerCase())) ?? '';
      }
      if (!entrypoint) {
        entrypoint = entries
          .filter((name) => DEFINITION_SUFFIXES.has(path.extname(name).toLowerCase()))
          .sort()
          .map((name) => path.join(source, name))[0] ?? '';
      }
    }
    if (!entrypoint || !(await isFile(entrypoint))) {
      throw new PublisherError(
        `The selected ${unitType} has no supported definition file.`,
        `invalid_${unitType}`,
      );
    }
  }

  const result = await candidate(unitType, source, workspace, entrypoint);
  if (unitType === 'plugin') result.children = await discoverPluginChildren(source);
  return result;
}

export function assertPublishTypeAvailable(unitType: string): void {
  if ((UNAVAILABLE_PUBLISH_TYPES as readonly string[]).includes(unitType)) {
    throw new PublisherError(
      `${unitType} publishing is not available yet. Taku Publisher currently accepts Skills only.`,
      'publish_type_not_available',
    );
  }
  if (!(SUPPORTED_TYPES as readonly string[]).includes(unitType)) {
    throw new PublisherError(`Unsupported publish type: ${unitType}`, 'unsupported_type');
  }
}

export async function discoverPluginChildren(pluginRoot: string): Promise<JsonObject[]> {
  const discovered = await discoverUnits(pluginRoot);
  return discovered
    .filter((entry) => !(entry.type === 'plugin' && entry.relative_path === '.'))
    .map((entry) => ({
      id: entry.id,
      type: entry.type,
      name: entry.name,
      relative_path: entry.relative_path,
    }));
}

async function discoverExplicit(source: string, workspace: string): Promise<JsonObject[]> {
  const candidates: JsonObject[] = [];
  if (await isDirectory(source)) {
    const manifest = await pluginManifest(source);
    if (manifest) candidates.push(await candidate('plugin', source, workspace, manifest));
    const skill = path.join(source, 'SKILL.md');
    if (await isFile(skill)) candidates.push(await candidate('skill', source, workspace, skill));
    if (!candidates.some((entry) => (SUPPORTED_TYPES as readonly string[]).includes(String(entry.type)))) {
      candidates.push(...await discoverUnits(source));
    }
  } else {
    const lower = path.basename(source).toLowerCase();
    const parent = path.basename(path.dirname(source)).toLowerCase();
    if (lower === 'skill.md') candidates.push(await candidate('skill', path.dirname(source), workspace, source));
    if (AGENT_FILE_NAMES.has(lower) || AGENT_PARENT_NAMES.has(parent)) {
      candidates.push(await candidate('agent', source, workspace, source));
    }
    if (ACTION_PARENT_NAMES.has(parent) || DEFINITION_SUFFIXES.has(path.extname(source).toLowerCase())) {
      candidates.push(await candidate('action', source, workspace, source));
    }
  }
  const unique = new Map<string, JsonObject>();
  for (const entry of candidates) unique.set(`${entry.type}\0${entry.path}`, entry);
  return [...unique.values()]
    .filter((entry) => (SUPPORTED_TYPES as readonly string[]).includes(String(entry.type)))
    .sort(compareCandidates);
}

async function visitWorkspace(
  workspace: string,
  current: string,
  depth: number,
  visitor: (directory: string, files: string[]) => Promise<void>,
): Promise<void> {
  const entries = await fsp.readdir(current, { withFileTypes: true });
  const files = entries.filter((entry) => entry.isFile()).map((entry) => entry.name).sort();
  await visitor(current, files);
  if (depth >= MAX_DISCOVERY_DEPTH) return;
  const directories = entries
    .filter((entry) => entry.isDirectory()
      && !entry.isSymbolicLink()
      && !EXCLUDED_DIR_NAMES.has(entry.name.toLowerCase())
      && !HOST_METADATA_DIR_NAMES.has(entry.name.toLowerCase()))
    .map((entry) => entry.name)
    .sort();
  for (const name of directories) {
    await visitWorkspace(workspace, path.join(current, name), depth + 1, visitor);
  }
}

async function pluginManifest(root: string): Promise<string | undefined> {
  for (const relative of PLUGIN_MANIFESTS) {
    const file = path.join(root, relative);
    if (await isFile(file)) return file;
  }
  return undefined;
}

async function candidate(
  unitType: string,
  sourceInput: string,
  workspace: string,
  entrypointInput: string,
): Promise<JsonObject> {
  const source = path.resolve(sourceInput);
  const entrypoint = path.resolve(entrypointInput);
  const relativePath = isWithin(source, workspace)
    ? normalizedRelative(source, workspace)
    : path.basename(source);
  const metadata = await readMetadata(entrypoint, source);
  const identifier = createHash('sha256')
    .update(`${unitType}\0${relativePath}\0${path.basename(entrypoint)}`)
    .digest('hex')
    .slice(0, 20);
  return {
    id: `${unitType}_${identifier}`,
    type: unitType,
    name: metadata.name,
    description: metadata.description,
    path: source,
    relative_path: relativePath,
    entrypoint,
    entrypoint_relative: (await isDirectory(source)) && isWithin(entrypoint, source)
      ? normalizedRelative(entrypoint, source)
      : path.basename(entrypoint),
  };
}

async function readMetadata(entrypoint: string, source: string): Promise<{ name: string; description: string }> {
  const sourceStat = await fsp.stat(source);
  const fallback = path.parse(sourceStat.isFile() ? path.basename(source) : source).name;
  if (path.extname(entrypoint).toLowerCase() === '.json') {
    try {
      const value = JSON.parse(await fsp.readFile(entrypoint, 'utf8')) as Record<string, unknown>;
      return {
        name: cleanText(value.name ?? value.title) || fallback,
        description: cleanText(value.description ?? value.summary),
      };
    } catch {
      // Fall through to text metadata.
    }
  }
  try {
    const text = (await fsp.readFile(entrypoint, 'utf8')).slice(0, 64_000);
    const frontmatter = parseFrontmatter(text);
    const heading = /^#\s+(.+?)\s*$/m.exec(text);
    return {
      name: cleanText(frontmatter.name ?? frontmatter.title) || heading?.[1]?.trim() || fallback,
      description: cleanText(frontmatter.description ?? frontmatter.summary),
    };
  } catch {
    return { name: fallback, description: '' };
  }
}

function parseFrontmatter(text: string): Record<string, string> {
  if (!text.startsWith('---')) return {};
  const end = text.indexOf('\n---', 3);
  if (end < 0) return {};
  const output: Record<string, string> = {};
  for (const line of text.slice(3, end).split(/\r?\n/)) {
    const match = /^([A-Za-z0-9_-]+):\s*["']?(.*?)["']?\s*$/.exec(line);
    if (match?.[1]) output[match[1]] = match[2] ?? '';
  }
  return output;
}

function cleanText(value: unknown): string {
  return typeof value === 'string' ? value.replace(/\s+/g, ' ').trim().slice(0, 500) : '';
}

function appendCandidate(candidates: JsonObject[], seen: Set<string>, entry: JsonObject): void {
  if (!(SUPPORTED_TYPES as readonly string[]).includes(String(entry.type))) return;
  const key = `${entry.type}\0${entry.path}`;
  if (seen.has(key)) return;
  seen.add(key);
  candidates.push(entry);
}

function compareCandidates(left: JsonObject, right: JsonObject): number {
  return `${left.type}\0${left.relative_path}\0${left.id}`.localeCompare(
    `${right.type}\0${right.relative_path}\0${right.id}`,
  );
}

async function isFile(candidatePath: string): Promise<boolean> {
  return fsp.stat(candidatePath).then((stat) => stat.isFile(), () => false);
}

async function isDirectory(candidatePath: string): Promise<boolean> {
  return fsp.stat(candidatePath).then((stat) => stat.isDirectory(), () => false);
}

async function isSymlink(candidatePath: string): Promise<boolean> {
  return fsp.lstat(candidatePath).then((stat) => stat.isSymbolicLink(), () => false);
}

async function canonicalPath(candidatePath: string): Promise<string> {
  const resolved = path.resolve(candidatePath);
  return fsp.realpath(resolved).catch(() => resolved);
}
